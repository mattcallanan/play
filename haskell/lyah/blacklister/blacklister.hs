-- ghci -XDeriveDataTypeable -XDeriveDataTypeable blacklister
import BlacklistData
import Data.List
import Data.List.Split -- for splitOn
import Control.Monad   -- for liftM
import System.CPUTime
import Text.Printf
import System.IO
import Control.Parallel.Strategies
import Control.Parallel

{-    
Storycard 1
-----------
- Using BlacklistData.hs, write "blacklister", a function that returns the customers that are blacklisted (their postcode is one of BlacklistData.blacklistedPostcodes).
- Use the Customer type exported from BlacklistData.
- Use testBlacklister below to check your results.
-}
instance Show Customer where
    show = intercalate "," . fromCustomer

ex = ["10100001","female","Alexandra","J","Kruse","29 Trelawney Street","BEN BUCKLER","NSW","2026","AU","AlexandraKruse@dodgit.com","Rae9ehee","(02) 9440 8706","Jenkins","3/11/1931","Visa","4.48589E+15","772","Oct-14","","1Z 706 396 06 2837 560 1","Microeconomist","Country Club Markets","PokerWatcher.com","A+","68.7","161"]
ex2 = toCustomer ex

toCustomer :: [String] -> Customer
toCustomer [idNumber, gender, givenName, middleInitial, surname, streetAddress, city, state, postCode, country, emailAddress, password, telephoneNumber, mothersMaiden, birthday, ccType, ccNumber, cvv2, ccExpires, nationalID, ups, occupation, company, domain, bloodType, kilograms, centimeters] = 
    Customer idNumber gender givenName middleInitial surname streetAddress city state postCode country emailAddress password telephoneNumber mothersMaiden birthday ccType ccNumber cvv2 ccExpires nationalID ups occupation company domain bloodType kilograms centimeters

--header :: Customer -> [String]
--header x = constrFields . toConstr $ x

fromCustomer :: Customer -> [String]
fromCustomer x = [idNumber x, gender x, givenName x, middleInitial x, surname x, streetAddress x, city x, state x, postCode x, country x, emailAddress x, password x, telephoneNumber x, mothersMaiden x, birthday x, ccType x, ccNumber x, cvv2 x, ccExpires x, nationalID x, ups x, occupation x, company x, domain x, bloodType x, kilograms x, centimeters x]

slurp :: [Customer]
slurp = map toCustomer customers

blacklisted :: [Customer] -> [Customer]
blacklisted = filter f where f Customer {postCode=pc} = (pc `elem` blacklistedPostcodes)

blacklister :: [String]
blacklister = map show $ blacklisted slurp
              --where f Customer {postCode=pc} = (pc `elem` blacklistedPostcodes)

testBlacklister = blacklister == expected


{-
Storycard 2
-----------
1. Add a "main" method to blacklister.hs that outputs blacklister to stdout (as lines of CSV).
2. Run "ghc blacklister"
3. Run "blacklister > blacklisted.csv"
4. Compare blacklisted.csv with expected.csv
-}

--main = mapM_ putStrLn $ blacklister
main = blacklisterPro

{-
Storycard 3
-----------
Write "blacklisterPro", an app that reads customers from "all_customers.csv" and has two forms of output:
1. Displays the IDs of all the blacklisted customers and a summary like so:
    nnnn customers blacklisted: [10100027,10100205,<etc>]
2. Writes the full records of blacklisted customers to blacklisted.csv (ordering is not important).  Use BlacklistData.columnNames for header row.
-}
-- liftM :: Monad m => (a1 -> r) -> m a1 -> m r
-- lines :: String -> [String]
-- readFile :: FilePath -> IO String
-- mapM_ :: Monad m => (a -> m b) -> [a] -> m ()

blacklisterPro :: IO b
blacklisterPro = do 
    start <- getCPUTime
    csvLines <- liftM lines (readFile "all_customers.csv")
    let allCustomers = map (toCustomer . (splitOn ",")) csvLines
    let blacklistedCustomers = blacklisted allCustomers
    let ids = map idNumber blacklistedCustomers
    putStrLn $ show (length blacklistedCustomers) ++ " customers blacklisted: " ++ (show ids)
    --mapM_ putStrLn (map show blacklistedCustomers)
    w <- openFile "blacklisted.csv" WriteMode
    hPutStrLn w (intercalate "," columnNames)
    mapM (hPutStrLn w) (map show blacklistedCustomers)
    hClose w
    end <- getCPUTime
    let diff = (fromIntegral (end - start)) / (10^12)
    printf "Computation time: %0.3f sec\n" (diff :: Double)

    {-
readCustomersText :: IO [String]
readCustomersText = do
    list <- liftM lines (readFile "all_customers.csv")
    return list

readCustomerTokens :: IO [[String]]
readCustomerTokens = (liftM (map $ splitOn ",") readCustomersText)

--readCustomers :: IO [Customer]
readCustomers = map (toCustomer) readCustomerTokens
-}
    
{-
Storycard 4
-----------
1. Write "blacklisterMapReduce", an app that retrofits "blacklisterPro" to process customers in at least 10 parallel threads/processes.
   - What difficulties are there in this approach?
   - What are the benefits of this approach?
-}
chunks :: [[Customer]]
chunks = (chunk 10000 slurp)

blacklistParallel = (parMap rseq) blacklisted chunks
