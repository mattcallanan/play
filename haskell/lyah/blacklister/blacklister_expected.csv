10100027,female,Dakota,A,McGuigan,83 Swanston Street,BOLANGUM,VIC,3381,AU,DakotaMcGuigan@mailinator.com,aeW4If2ie,(03) 5323 1922,Liebe,3/25/1954,Visa,4.53935E+15,642,Mar-16,,1Z 641 671 96 8044 383 0,Agricultural laborer,Sambo's,DishSuperstore.com,A+,62.2,169
10100205,male,Hunter,S,Fizelle,29 Pelican Road,TANAMI,WA,6770,AU,HunterFizelle@spambob.com,vooChoC3eeTh,(08) 9045 0081,Scollen,1/13/1968,MasterCard,5.29567E+15,282,Nov-15,,1Z 913 1E7 68 9300 295 5,Linemen,Terra Nova Garden Services,TrustApproval.com,O+,67.1,188
10100218,female,Gabriella,A,Godson,13 Old Tenterfield Road,UPPER DUCK CREEK,NSW,2469,AU,GabriellaGodson@trashymail.com,woob3Aiviwi,(02) 6720 4382,Clucas,4/05/1962,MasterCard,5.46177E+15,745,Aug-17,,1Z A28 060 13 6543 464 7,Clerical specialist,Crown Books,ResidentialStore.com,O+,77.4,153
10100288,male,Connor,I,De Mole,9 Chatsworth Road,DEEP CREEK,NSW,2469,AU,ConnorDeMole@trashymail.com,Op7naiTh2ie,(02) 6793 7863,Byles,10/12/1992,Visa,4.48544E+15,685,Apr-15,,1Z 917 971 46 5754 874 2,Electronic data processor,Honest Air Group,TrueLiberals.com,O+,75.3,176
10100311,female,Alica,L,Evans,67 Old Tenterfield Road,THERESA CREEK,NSW,2469,AU,AlicaEvans@spambob.com,ThohGo1oo,(02) 6736 5710,Ellen,2/12/1947,MasterCard,5.57296E+15,898,Aug-16,,1Z A56 749 27 9023 386 3,Water transportation master,Landskip Yard Care,KosherMobile.com,A+,75,154
10100325,female,Stephanie,J,Bradford,13 Darwinia Loop,MCBEATH,WA,6770,AU,StephanieBradford@trashymail.com,leeQuoob0Pai,(08) 9027 9832,Brient,7/24/1969,MasterCard,5.50846E+15,204,Jul-16,,1Z V21 81V 12 7103 913 2,Explosives worker,Druther's,DomainDebate.com,B+,61.9,173
10100420,male,Cameron,A,Wickham,56 Old Tenterfield Road,OLD BONALBO,NSW,2469,AU,CameronWickham@spambob.com,ohfeegh0Th,(02) 6772 4721,Baines,10/29/1930,Visa,4.55616E+15,280,Jul-13,,1Z V48 175 22 7452 254 5,Job analysis manager,Pomeroy's,LimitCap.com,A+,75.6,172
10100425,female,Lauren,R,Allwood,81 Ridge Road,GOODWOOD,QLD,4660,AU,LaurenAllwood@mailinator.com,phoo2DoeKee,(07) 3337 5352,Frankland,5/27/1956,MasterCard,5.58938E+15,386,Jan-13,,1Z 3Y0 519 20 1806 390 0,Lathe and turning machine tool operator,Zany Brainy,AudioFail.com,O+,53.6,174
10100437,female,Lilian,C,Crooke,54 Old Tenterfield Road,THERESA CREEK,NSW,2469,AU,LilianCrooke@spambob.com,Tahz7ie7tee6,(02) 6730 4811,Telfer,2/20/1981,MasterCard,5.13245E+15,387,Jun-17,,1Z 073 705 93 0249 976 3,Faller,Dream Home Improvements,FlyFunds.com,O+,70,173
10100491,female,Bethany,J,Macnaghten,38 Old Tenterfield Road,MUMMULGUM,NSW,2469,AU,BethanyMacnaghten@dodgit.com,thaezie1Fi7ee,(02) 6774 3277,Bidencope,9/13/1976,MasterCard,5.24111E+15,230,Dec-13,,1Z 801 E12 28 3238 616 5,Avionics technician,Father & Son,ToySearcher.com,A+,68.8,170
10100533,male,Anthony,L,Vincent,59 Cunningham Street,WESTMAR,QLD,4422,AU,AnthonyVincent@spambob.com,iephahL6g,(07) 4595 9332,Warby,8/21/1955,MasterCard,5.52531E+15,743,Jun-17,,1Z 161 668 52 2955 709 2,Mental health counselor,Fedco,ColdBerg.com,O+,79.9,186
10100580,female,Indiana,C,Gant,35 Mendooran Road,BOOTHENBA,NSW,2830,AU,IndianaGant@mailinator.com,ohNu0ahvoo,(02) 4041 2222,Arnold,10/30/1991,MasterCard,5.53471E+15,619,Jul-16,,1Z 81F W53 81 0722 987 4,Biological scientist,Flipside Records,ChessWhiz.com,O+,72.7,170
10100810,female,Jorja,N,Weindorfer,77 Chatsworth Road,BULLDOG,NSW,2469,AU,JorjaWeindorfer@pookmail.com,ahxaiT5ya,(02) 6773 7932,Maudsley,8/14/1978,Visa,4.53946E+15,541,Aug-14,,1Z F70 784 02 2741 946 1,Actuarie,Muscle Factory,ItSummary.com,O+,61.6,166
10100831,female,Mary,C,Carr-Glyn,59 Chatsworth Road,BEAN CREEK,NSW,2469,AU,MaryCarr-Glyn@pookmail.com,Roowohng6ei,(02) 6740 5625,Arscott,2/05/1935,MasterCard,5.14805E+15,542,Jun-17,,1Z 865 929 69 0015 628 0,Traffic clerk,Integra Design,CoachConference.com,A-,104.2,160
10100865,female,Ellie,C,Tebbutt,83 Lapko Road,BORDEN,WA,6338,AU,EllieTebbutt@spambob.com,uc7Oog1joh,(08) 9021 1438,Fletcher,12/24/1978,MasterCard,5.30744E+15,678,Jun-16,,1Z 492 628 69 6322 957 9,Crushing grinding and polishing machine operator,Good Times,IRSCoupons.com,B+,76.5,171
10100971,male,Patrick,I,McCue,29 Wharf St,MOUNT WHITE,NSW,2250,AU,PatrickMcCue@trashymail.com,UuHuo6er9,(02) 9654 8486,Jorgenson,11/10/1934,MasterCard,5.47886E+15,712,Sep-14,,1Z 317 072 13 9367 283 1,Hand packager,Eli Moore Inc,WirelessCopier.com,B-,71.3,176
